lib/spack/llnl/util/filesystem.py

```diff
@@ -2838,6 +2839,28 @@ def temporary_dir(
         remove_directory_contents(tmp_dir)
 
 
+@contextmanager
+def edit_in_place_through_temporary_file(file_path: str) -> Generator[str, None, None]:
+    """Context manager for modifying ``file_path`` in place, preserving its inode and hardlinks,
+    for functions or external tools that do not support in-place editing. Notice that this function
+    is unsafe in that it returns a path instead of a file descriptor, but this is by design, since
+    we assume an external tool will modify the file at the temporary path by creating a new
+    inode."""
+    tmp_fd, tmp_path = tempfile.mkstemp(
+        dir=os.path.dirname(file_path), prefix=f"{os.path.basename(file_path)}."
+    )
+    try:
+        with open(file_path, "rb") as f, os.fdopen(tmp_fd, "wb", closefd=False) as g:
+            shutil.copyfileobj(f, g)
+            g.flush()
```

Comment:
Does this need `copyfileobj`, or would `shutil.copyfile` be enough?

Reply:
I thought I'd use the file descriptor because it's safe, but in reality that doesn't matter since the file is replaced later. Looks like copyfile is a direct syscall (fcopyfile) so to be preferred yeah, not sure why copyfileobj doesn't have a fast path...

Reply:
Ok! 

I just confirmed that with the latest version of this PR and switching to `copyfile` the hardlinks remain intact in my test!

Reply:
with a4bda74?

Reply:
Yes, exactly. Although I pull it as a patch on top of #47370 because the installation fails otherwise.

Reply:
Right, I missed all those pings there.